File: FindInJson/FindInJson/JsonNode.cpp
#include "JsonNode.h"

JsonNode::JsonNode()
{
}

JsonNode::JsonNode(const std::string& key, JsonNode* parent = NULL)
{
	_key = key;
	_parent = parent;
	if (parent != NULL)
		parent->_childs.push_back(this);
}

const std::string JsonNode::get_key()
{
	return _key;
}

std::vector<JsonNode*> JsonNode::get_childs()
{
	return _childs;
}

JsonNode* JsonNode::get_parent()
{
	return _parent;
}

void JsonNode::set_key(const std::string& key)
{
	_key = key;
}

void JsonNode::add_child(JsonNode* child)
{
	_childs.push_back(child);
}

void JsonNode::set_node_type(NodeType type)
{
	_type = type;
}

std::string JsonNode::to_string(int tubs)
{
	std::string res;

	if (_type == NodeType::OBJECT)
		res += "{";

	tubs++;

	for (int i = 0; i < _childs.size(); i++)
	{
		auto node = _childs[i];

		if (node->_childs.size() != 0)
			res += "\n" + add_tubs(tubs);

		if (node->_type != NodeType::NUM)
			res += "\"";

		res += node->get_key();

		if (node->_type != NodeType::NUM)
			res += "\"";

		if (node->_childs.size() != 0)
			res += ": " + node->to_string(tubs);

		if (i <= _childs.size() - 2 && _childs.size() > 1)
			res += ",";
	}

	if (_type == NodeType::OBJECT)
		res += "\n" + add_tubs(tubs - 1) + "}";

	return res;
}

JsonNode::~JsonNode()
{
}

std::string JsonNode::add_tubs(int tubs)
{
	std::string str;

	for (int i = 0; i < tubs; i++)
		str += '\t';

	return str;
}

File: FindInJson_C#/FindInJson/JsonNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FindInJson
{
    enum NodeType {NUM, STRING, OBJECT, NONE};
    internal class JsonNode
    {
        string _key;
        List<JsonNode> _childs;
        JsonNode _parent = null;
        NodeType _type = NodeType.NONE;

        public string Key 
        { 
            get { return _key; } 
            set { _key = value; } 
        }
        public List<JsonNode> Childs { get { return _childs; } }
        public NodeType Type { get { return _type; } set { _type = value; } }
        public JsonNode Parent { get { return _parent; } }

        public JsonNode() { }
        public JsonNode(string key, JsonNode parent) 
        {
            _key = key;
            _parent = parent;
            if (parent != null)
                parent.AddChild(this);
        }
        public void AddChild(JsonNode child)
        {
            if (_childs == null)
                _childs = new List<JsonNode>();
            _childs.Add(child);
        }
        override public string ToString()
        {
            return ToString(0);
        }
        string ToString(int tabs)
        {
            var res = "";

            tabs++;

            for (int i = 0; i < _childs.Count; i++)
            {
                var node = _childs[i];

                if (node._childs != null)
                    res += $"\n{AddTabs(tabs)}";

                if (node._type != NodeType.NUM)
                    res += $"\"{node._key}\"";
                else
                    res += node._key;

                if (node._childs != null)
                    res += $": {node.ToString(tabs)}";

                if (i <= _childs.Count - 2 && _childs.Count > 1)
                    res += ",";
            }

            if (_type == NodeType.OBJECT)
                res = $"{{{res}\n{AddTabs(tabs - 1)}}}";

            return res;
        }
        string AddTabs(int tabs)
        {
            var str = "";

            for (int i = 0; i < tabs; i++)
                str += '\t';

            return str;
        }
    }
}
